### DISCUSSION TITLE
How to access `auth` schema via PostgREST?

### DISCUSSION THREAD

**Discussion by liaujianjie** [supabase/supabase#3599] [Questions]
I'm looking to integrate a SSO+SCIM SaaS into my app. So I'll have to manually handle user creation and access token minting on one of my own custom endpoints after a successful SSO sign in. Since my endpoints are serverless, I'd like my endpoint implementations to use the PostgREST server hosted by Supabase so I can use its existing Postgres connection pool. Yet, it seems that the Supabase-hosted PostgREST instance only allows access to the `public` and `storage` schema when I tried to access the endpoint with and `apikey` of `role: "service_role"`.

Is there a way to allow access to the `auth` schema via PostgREST?

Or is there an better workflow to handle user creation and access token minting for 3rd-party auth?

**Comment by steve-chavez** [comment]
Note that the authentication logic is handled by [gotrue](https://github.com/supabase/gotrue), not PostgREST.

> Is there a way to allow access to the auth schema via PostgREST?

You could create a `public.users` view of the `auth.users` table. That being said, I'm not sure if you can insert rows into `auth.users` and not mess with gotrue logic.

> Or is there an better workflow to handle user creation and access token minting for 3rd-party auth?

@kangmingtay Might have a better idea on how to do this.

**Reply by liaujianjie** [reply]
> Note that the authentication logic is handled by [gotrue](https://github.com/supabase/gotrue), not PostgREST.

Yup, I'm aware. I was just hoping to also access the `auth` schema via PostgREST so that I can manually handle user creation.

> You could create a `public.users` view of the `auth.users` table. That being said, I'm not sure if you can insert rows into `auth.users` and not mess with gotrue logic.

I'm fine with getting my hands dirty with replicating some GoTrue logic on my end. My only concerns are:
1. The completeness of my implementation. For example, should I be also writing to `auth.audit_log_entries`?
2. The brittleness of having to stay in sync with auth logic changes from GoTrue.

**Comment by liaujianjie** [comment]
An alternative solution @bnjmnt4n came up with was to write Postgres stored procedures to handle user creation and refresh token minting. For example, I could mint a token with:

```sql
-- Reimplementation of the same cryptographically secure token generator from Supabase GoTrue.
-- Source: https://github.com/supabase/gotrue/blob/6628f81d11d99daf8fd73346df34e328976a944b/crypto/crypto.go#L16-L23
create or replace function generate_secure_token()
returns varchar(255) as $$
  declare
    token text;
  begin

  token := encode(gen_random_bytes(16), 'base64');
  token := rtrim(token, '=');
  token := replace(token, '/', '_');
  token := replace(token, '+', '-');

  return token::varchar(255);
end;
$$ language plpgsql
set search_path = extensions;
```

So far, I've only gathered that I'd need to replicate these logic into Postgres stored procedures:
1. Insertion of `auth.users` for new user creation
2. Selection of `auth.users` to check for existing user
3. Minting of `auth.refresh_tokens` for a new login session
Am I missing anything? Would be nifty if we had a sequence diagram for Supabase GoTrue auth! I would contribute but I'm not sure I fully understand how GoTrue works.

Then, to secure these stored procedures, I can add role checks:

```sql
  if auth.role() is null then
    raise exception 'Function invoker must have a role.';
  end if;
  if auth.role() <> 'service_role' then
    raise exception 'Role has no permission to execute function: %', auth.role();
  end if;
```
